src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp

```diff
@@ -15,751 +15,878 @@
+                            aliveBosses.push_back(steelbreaker);
 
-                    if (me->HasUnitState(UNIT_STATE_CASTING))
-                        return;
-                }
+                    if (Creature* brundir = _instance->GetCreature(DATA_STORMCALLER_BRUNDIR))
+                        if (brundir->IsAlive())
+                            aliveBosses.push_back(brundir);
 
-                DoMeleeAttackIfReady();
-            }
-        };
+                    if (Creature* target = Trinity::Containers::SelectRandomContainerElement(aliveBosses))
+                        DoCast(target, SPELL_RUNE_OF_POWER);
 
-        CreatureAI* GetAI(Creature* creature) const override
-        {
-            return GetUlduarAI<boss_runemaster_molgeimAI>(creature);
+                    _events.Repeat(32s, 47s);
+                    break;
+                }
+                case EVENT_SHIELD_OF_RUNES:
+                    DoCastSelf(SPELL_SHIELD_OF_RUNES);
+                    _events.Repeat(32s, 48s);
+                    break;
+                case EVENT_RUNE_OF_DEATH:
+                    Talk(SAY_MOLGEIM_RUNE_DEATH);
+                    if (Unit* target = SelectTarget(SelectTargetMethod::Random))
+                        DoCast(target, SPELL_RUNE_OF_DEATH);
+                    _events.Repeat(47s, 54s);
+                    break;
+                case EVENT_RUNE_OF_SUMMONING:
+                    Talk(SAY_MOLGEIM_SUMMON);
+                    _summons.RemoveNotExisting(); // Keeping the summon list clean
+                    if (Unit* target = SelectTarget(SelectTargetMethod::Random))
+                        DoCast(target, SPELL_RUNE_OF_SUMMONING);
+                    _events.Repeat(23s, 33s);
+                    break;
+                default:
+                    break;
+            }
+
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
         }
+
+        DoMeleeAttackIfReady();
+    }
+private:
+    EventMap _events;
+    InstanceScript* _instance;
+    SummonList _summons;
 };
 
-class boss_stormcaller_brundir : public CreatureScript
+struct boss_stormcaller_brundir : public ScriptedAI
 {
-    public:
-        boss_stormcaller_brundir() : CreatureScript("boss_stormcaller_brundir") { }
-
-        struct boss_stormcaller_brundirAI : public BossAI
+    boss_stormcaller_brundir(Creature* creature) : ScriptedAI(creature), _instance(nullptr), _summons(me) { }
+
+    void InitializeAI() override
+    {
+        // For dynamic linking compatability
+        _instance = me->GetInstanceScript();
+        SetBoundary(&ArenaBoundaries);
+    }
+
+    void JustAppeared() override
+    {
+        me->ApplySpellImmune(0, IMMUNITY_MECHANIC, MECHANIC_INTERRUPT, false);  // Should be interruptable unless overridden by spell (Overload)
+        me->ApplySpellImmune(0, IMMUNITY_MECHANIC, MECHANIC_STUN, false);       // Reset immumity, Brundir should be stunnable by default
```

Comment:
those immunities should be removed from the script, also below PHASE_THREE stun immunitie

and we should remove from DB stun immunity
```sql
-- Stormcaller Brundir: remove stun immunity
UPDATE `creature_template` SET `mechanic_immune_mask`=`mechanic_immune_mask` &~(2048) WHERE `entry` IN (32857,33694);
```
Overload and Stormshield auras gives proper immunities to the boss